### DISCUSSION TITLE
Related entities are inserted multiple times

### DISCUSSION THREAD

**Discussion by justkey007** [mikro-orm/mikro-orm#3677] [Q&A] [answered]
**Describe the bug**
Entities that are already persisted and are related to a new entity that is not yet persisted are recreated when saving new entities and this generates a `UniqueConstraintViolationException`.

**Stack trace**
```
UniqueConstraintViolationException: insert into "account" ("id", "created_at", "updated_at", "email", "confirmed", "role", "password")
values ('05148d45-8639-42ec-a152-44743c281b1c', '2022-11-01T08:46:42.321Z', NULL, 'xxxx@xxxxx.com', true, 'CUSTOMER',
'$argon2id$v=19$m=4096,t=3,p=1$mrFYDR5isGanNDa7lwsrZA$xlPml/+C4BgnTMgpmVrIcAymg21j/+kFy0klOyhAt+k') returning "id",
"version", "confirmed", "enable" - duplicate key value violates unique constraint "account_pkey"
```

**To Reproduce**
```typescript
const customer = this.customerRepository.create(data);
const account = await this.accountRepository.nativeInsert(accountData);
customer.account = account.id;
await this.customerRepository.persistAndFlush(customer); // throw UniqueConstraintViolationException: insert into "account" ...
```
When I try to clear the context of the EM I no longer have an exception 👇.
```typescript
const customer = this.customerRepository.create(data);
const account = await this.accountRepository.nativeInsert(accountData);
customer.account = account.id;

this.orm.em.clear();
await this.customerRepository.persistAndFlush(customer);
```

**Expected behavior**
The expected behavior is the persistence of the customer without throwing an exception as in version 5.4.2.

**Additional context**

**Versions**
5.5.0

| Dependency | Version |
| - | - |
| node | v14.19.3 |
| typescript | v4.8.4 |
| mikro-orm | 5.5.0 |
| your-driver | @mikro-orm/postgresql |

**Comment by B4nan** [comment, marked as answer]
Please provide complete (but minimal) reproduction, I have no idea what you are actually doing from that short snippet. You are using some service method there, how can I know what it does?

Ideally, it should be a simple script, I don't want to browse through many files. Check existing [regression tests](https://github.com/mikro-orm/mikro-orm/tree/master/tests/issues) for inspiration, or use this [stackblitz.com/edit/mikro-orm-repro](https://stackblitz.com/edit/mikro-orm-repro) as a base for it.

**Reply by justkey007** [reply]
@B4nan
Please, there is an important detail that I forgot to give you. In AccountsService I used the `nativeInsert` method for creating the account.
For the unit of work it is a new entity to create
```typescript
await this.accountRepository.nativeInsert(account);
```

**Reply by B4nan** [reply]
`nativeInsert` is not affecting the context anyhow. It is the same as firing insert query directly. You are clearly mixing the two approaches.

**Reply by B4nan** [reply]
> customer.account = account.id;

Even this looks wrong, your `Custom.account` property is what? This looks like you are using `mapToPk`?

As said above, please provide complete repro, there are too many blank spots. Can't help without seeing what you are doing.

**Reply by justkey007** [reply]
But I didn't get this error with version 5.4.2. It was immediately that I updated to v5.5.0 that my tests no longer pass.

**Reply by B4nan** [reply]
That's irrelevant, the ORM tests also didn't change. You were clearly doing something wrong before, or depending on some undefined behavior.

**Reply by justkey007** [reply]
It's possible, as soon as I have news after my analyzes I'll get back to you. Thank you very much for your responsiveness.

**Reply by justkey007** [reply]
By adding the `persistOnCreate` option to `false` in the MikroOrm configuration the problem disappeared and my tests were successfully executed. Thanks

**Reply by B4nan** [reply]
That still means you are doing something wrong most probably. I'll be happy to tell you more once I can see what you are actually doing.

(or it may as well be some internal bug, again, not possible to say without seeing complete repro)

**Reply by B4nan** [reply]
Like creating managed entities via `em.create()` without `managed: true` in the options, that's one example that could lead to such behavior.

**Reply by B4nan** [reply]
Up until now, you haven't shared the entity definitions, let's start with that. I was asking for what the `Customer.account` is mainly, as it feels wrong what you are doing in there. `em.nativeInsert` does not return entity instance, so I don't think there is gonna be any `id` property on the object it returns - it should not return _any object_, it should return the PK, not entity.

**Reply by justkey007** [reply]
I will try to reproduce the issue in a repo and provide it to you. Thanks

**Reply by justkey007** [reply]
@B4nan 
https://github.com/justkey007/mikroorm-3677

By removing the comment from this line the test will pass successfully. Thanks
https://github.com/justkey007/mikroorm-3677/blob/44912b52e15710538643d4b7f24e63d752a7251c/src/config/mikro-orm-module.config.ts#L16

**Reply by B4nan** [reply]
[This](https://github.com/justkey007/mikroorm-3677/blob/44912b52e15710538643d4b7f24e63d752a7251c/src/modules/customers/entities/customer.entity.ts#L17) is indeed wrong, it always needs to be entity instance, you cant mix PKs with entity instances like that. 

```diff
-account: Account | string;
+account: Account;
```

If you want more help, please provide minimal repro as I asked for, this is a complex web app, and I don't have the capacity to help you with such. I also saw you are explicitly allowing global context, that's a huge red flag, it opens your app to memory leaks and instable API response issues.

**Reply by justkey007** [reply]
Thank you for your comments, I have just corrected them and yet it did not solve the problem. The repo provided is not complex I wrote it for the situation and there are only two entities in this repo (Account and Customer).

To run the tests after running the database:
`npm run test:e2e`

**Reply by B4nan** [reply]
> The repo provided is not complex 

It is complex, very complex. Check the links to regression tests or the stackblitz I gave you - that is what a minimal but complete reproduction looks like. 

You are missing my point. To reproduce something you usually need one or two entities, with very few properties, and that's it. Not web apps, not migrations, not complex dot env configurations, not 3 different layers (controller, service, module) in many nested folders, nothing from your _very complex_ repro you gave me. That only adds clutter and costs me a lot of time. I am not interested in paying that time for free, because _you_ want it 🤷 I am here to help with ORM issuses, not with your web apps. If you want me to help you with your web app, you can pay me, otherwise, you will need to provide minimal reproductions without all that clutter. 

edit: Also note that I already told you what you are doing wrong. that `create` service method is your issue, you are mixing UoW approach with native inserts. This was never expected to work. You might wanna try `em.upsert` added in v5.5, it is doing something similar but works with the UoW natively.

**Reply by justkey007** [reply]
The objective is not to make you work for free. I have already contributed to this repo once and my goal is to do so again.

I have just updated the repo for the reprodutcion of the error still on the same link but it must be cloned again because I have reset it in order to provide you with a minimal code.

The test file is [here](https://github.com/justkey007/mikroorm-3677/blob/main/tests/customer.spec.ts)

**Reply by justkey007** [reply]
> edit: Also note that I already told you what you are doing wrong. that create service method is your issue, you are mixing UoW approach with native inserts. This was never expected to work. You might wanna try em.upsert added in v5.5, it is doing something similar but works with the UoW natively.

It works with the `em.upsert` method. We can close the discussion and thank you very much for the assistance.

**Reply by B4nan** [reply]
As said above, you can't do this:

```ts
// in v5.5 this automatically persists the newly created entity, so the EM is aware of it and will try to insert it on flush
const account = accountsRepository.create(data); 
await accountsRepository.nativeInsert(account);
return account;
```

It will always result in insert statement via `flush`, hence fail with the uniqueness error, as you already create the entity in database via `nativeInsert`. `nativeInsert` has no effect on UoW, its a shortcut for QB basically. 

You can indeed revert to the pre-5.5 behaviour as you already know. That does not mean the code is correct, I would still suggest using `em.upsert` instead, it will run `insert on conflict update` query if the entity does not have a PK set. If the entity is loaded in context, it will do `em.assign()` on it, so next flush operation will issue update query. Sounds like exactly what you are trying to do.

Or stop fighting the UoW, you don't need to know the PKs before you flush:

```ts
const customer = em.create(Customer, data);
const account = em.create(Account, { ...data, role: 'CUSTOMER' });
customer.account = account;
// if you disable the `persistOnCreate`, you will also need to call `em.persist()` on at least one of the entities
// em.persist(customer);
await em.flush();
return customer;
```

**Reply by justkey007** [reply]
Very clear, thank you
